initial commit

ytv opened the pull request.

ytv committed:
initial commit

darrenvng commented on main.html:
```diff
@@ -0,0 +1,69 @@
+    <meta name="viewport" content="width = device-width, initial-scale = 1">
+    <link rel="stylesheet" href="public/css/app.css">
+    <script src = "https://ajax.googleapis.com/ajax/libs/angularjs/1.4.9/angular.min.js"></script>
+    <script src="public/js/app.js"></script>
```
Keep the file names consistent, lots of developers are picky about this, app.ctrl.js

ytv replied:
You're referring to the capitalization here?

darrenvng commented on main.html:
```diff
@@ -0,0 +1,69 @@
+<!DOCTYPE html>
```
Branch names aren't variables so you shouldn't name it firstCommit, use a hyphenated structure like 'initial-commit'.  Also, you can start making better commit messages since your projects are more complicated now, a good way is a short descriptive title, line separator, then short summary.  If its a small commit, you can just have a short title.

Initial commit

Initial implementation of base angular views and controllers, added css styling, set up single page router.

ytv replied:
noted for next time

darrenvng commented on main.html:
```diff
@@ -0,0 +1,69 @@
+
+                <div class="row3 sm-label">
+                    <span class="row3__label__off">OFF</span>
+                    <span class="row3__btn-border">
```
Its pretty important to make sure your user experience is idiot proof.  Make the default game state 'on', worst thing is that when someone tries to play by clicking start without turning the game on and assumes its a bug.  Start button should probably be green colored.  Maybe you can also add a simple alert message for when user tries to start a game but the app is off.

ytv replied:
Good point.  I changed it so the game started on the 'on' position.  However, I left the start button red because that's how the original Simon game looked.  I wanted it to resemble the actual toy as much as possible.

darrenvng commented on main.html:
```diff
@@ -0,0 +1,69 @@
+                <div class="inner-circle">
+                <div class="inner-circle__label"><b>simon</b></div>
+
+                <div class="row1">
```
Unless you have a really, really good reason, don't use underscores for css classes, use hyphen separated words.  You also remove unnecessary classes like row1__screen_txt by just using a span inside your parent div and referring to the span through the parent class.  Also, don't use any numbers in your names like row1, row2.  If you wanted to change the positioning of these views like moving the buttons elsewhere, your names would no longer make sense.  Use a separate css class name instead of row2, row3.  Look up css naming conventions for proper ways to name things in different situations.

ytv replied:
I was using the BEM convention, but I'll think about the class names

ytv replied:
Ok, I got rid of the rows and replaced the underscores with hyphens.  Also reorganized the css name groupings and edited other css class names to make it more functional instead of presentational.

darrenvng commented on main.html:
```diff
@@ -0,0 +1,69 @@
+                    ng-style="blue"
+                    ng-mousedown="lightUp(4)"
+                    ng-mouseup="turnOff(4); passUserInput(4)">
+            </button>
```
middle-circle isn't a good name cause if you move your circle to the top, now you have to change your name for it to make sense again.

ytv replied:
Renamed to "game-ctrl-border" as that was its only "function".

darrenvng commented on public/css/app.css:
```diff
@@ -0,0 +1,310 @@
+/* container */
+
+.container {
+    position: relative;
```
definitely fix indentation so its nice and clean looking

ytv replied:
But I feel the indentation under "position" makes it easier to make changes later, because left/top/transform are all very pertinent to repositioning elements which happens frequently with a project like this.

darrenvng commented on public/css/app.css:
```diff
@@ -0,0 +1,310 @@
+}
+
+ /* inner circle */
+
```
If the inner-circle is really just the game controls, then you can probably call it something like that.  You don't want to tie your code to a visual position as much as possible, since positions can change.  Start thinking like this now, the less dependencies your code has the better.  Choosing a good balance between being flexible and being descriptive.

ytv replied:
renamed to game-cntrl

darrenvng commented on public/css/app.css:
```diff
@@ -0,0 +1,310 @@
+    left: 17.5%;
+    top: 17.5%;
+    width: 65% !important;
+}
```
Don't use the native browser interpretations of colors like 'red'.  Choose the actual hex code.  Better to use #000000 than black.

ytv replied:
Ok, changed it.

darrenvng commented on public/css/app.css:
```diff
@@ -0,0 +1,310 @@
+}
+
+/* inner circle row3 */
+
```
row3 implies that it is the same as row2 and row1, just at the bottom.  If they are the same, just call it row, and conserve as many css names as possible. 'row .span' instead of '.row-title'

ytv replied:
The row classes have been removed altogether.  I only used them to position "groups" of elements, but you right in that it wasn't very necessary.

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+App.controller('App.Ctrl', function($scope, $timeout, AppService) {
+    const colorCode = {1:"green", 2: "red", 3: "yellow", 4: "blue"};
+    const speed = 500; // in milliseconds
+    var on = false;
```
Start using some emca6 syntax.  In general using 'let' is superior to 'var', read about the differences as you will probably be tested on stuff like that.  When you have a ton of variables like this, its definitely a sign that you can change your app logic somewhat to get rid of some of them.  Even then, if you have consecutive var declarations, use one var, and separate the rest with commas, with one variable per line, indenting all but the first one.

ytv replied:
Changed all var's to lets, with a block let declaration at the top.

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+        }
+    };
+
+    turnOnAllLights = function() {
```
DRY principle, don't repeat yourself.  If your app grows in complexity, this won't scale.  Use a loop!!

ytv replied:
Done

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+    playSound = function(n) {
+        sound[n].play();
+    };
+
```
Use a loop instead of repeating code.  Even then, this wouldn't scale.  Even if your app audio grows only linearly (you start adding different music packages that user selects), you would have all this extra code in your app controller managing all this.  Even worse, these links are hard coded.  Once you have more than 10 links, it starts becoming unmanageable.  Create an angular config and include a config section for audio to store the link names.  Add an audio service '/js/services/audio.service.js' with some methods to get stuff.  Like getWinAudioElement(), which returns an element with the src already set.

ytv replied:
I created an audio service ('/js/services/audio.service.js') to completely the audio, removed the hardcode, and utilized a loop for scaling as well as a factory like function to create audio elements.

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+    };
+
+    wrongInput = function() {
+        buzzer.play();
```
Minor, but screen as a name is ambiguous, prob call it textDisplay or something.

ytv replied:
Done

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+            else resetToRound1(true);
+        }, 1500);
+    };
+
```
This is a good example of where a comment would have a high clarification to word ratio.  Probably only 1/100 programmers can look at toggleLights and know how it works after one pass through.  Probably 1-2 lines for comments will suffice.

ytv replied:
Done

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+App.controller('App.Ctrl', function($scope, $timeout, AppService) {
```
Definitely use the array method of including services in angular, if this code gets minified/compressed, it won't work.
someModule.controller('MyController', ['$scope', 'dep1', 'dep2', function($scope, dep1, dep2) {....
Do this for all of angular, including services, etc.

darrenvng replied:
Add a very short comment at the top describing the rules of the game or how to play.  A developer who knows the game will find it easier to navigate your code, many people will never have heard of 'Simon Says'.

ytv replied:
Added the array method for angular code and rules at the top of the game.

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+App.controller('App.Ctrl', function($scope, $timeout, AppService) {
+    const colorCode = {1:"green", 2: "red", 3: "yellow", 4: "blue"};
```
This is a personal preference for me, but I find it makes your code easier to navigate.  Any constants that won't change I usually name them with all aps and underscore styling.  Like COLOR_CODE, so its crystal clear anywhere in the code that this is a constant.  Also, I think its better to use COLOR_CODE.GREEN = 'green' instead of COLOR_CODE[1] = green since green should be green regardless of positioning.  Try to find a way to not bind your color constants with your app logic.

ytv replied:
I changed colorCode to COLOR_CODE and capitalized all other constant variables.  However I kept the code numbers for scaling purposes and in case I wanted to change the colors or their order in the future.

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+    var round = 1;
+    var turnOffTimer;
+    var recursiveTimer;
+    var timer;
```
{green:{}, red:{} ...} is better as it doesn't imply any type of positioning.  Green shouldn't care who, what or where red is.  It just knows only about himself.

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+
+    $scope.toggleOnOff = function() {
+        if(on === true) {
+            on = false;
```
Definitely remove hard coded css attributes from your controllers.  Angular has a lot of boiler plate code to help you handle display logic.  https://docs.angularjs.org/api/ng/directive/ngClass.  All you need is a scope variable, ie btnOn = false.  Then you leverage angular ngclass expressions, and it will dynamically toggle your css class (always use a class rather than have hard coded css attributes) based on your expression and scope variables.  Try to move all the display logic out of your app.ctrl.  Angular html has access to the scope, you need only initialize your scope variable and let the angular html directives like ngClick change the value of those scope display variables.  In fact, it should only be a small section near the top of app.ctrl where you initialize those scope params with a comment: // initialize scope display variables

ytv replied:
I created a scope variable $scope.on and utilized the ng-class directive to change the css classes of the on/off buttons and strict button.

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+            $scope.btn__on = {"background-color": "#8A96B6"};
+        }
+    };
+
```
Just like the above, use angular directives and you don't need this function anymore.  fyi ... for toggling boolean values, its easier to simple say booleanValue = !booleanValue (do this in the html, and toggle the css classes in the html accordingly)

ytv replied:
Used your booleanValue = !booleanValue expression in ng-click to toggle the strict button.  I kept the $scope.toggleOnOff function though because I need other functions to execute (i.e. cancelAllTimers(), turnOffAllLights() ).

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+         if(on === true && lock === false) {
+             cancelAllTimers();
+             turnOffAllLights();
+             var result = AppService.addUserInput(n, round);
```
Even if you don't have a default state, add the switch default and just break on it.

ytv replied:
done

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+             turnOffAllLights();
+             var result = AppService.addUserInput(n, round);
+             switch(result) {
+                 // user input is incorrect
```
Using integer values to reflect state make it hard for people to understand your program.  What was 1 again?  Declare and use constants for stuff like this.  STATE.WIN = 'win'.  case STATE.WIN:

ytv replied:
Changed to 

const STATE = {WIN: 'win', WRONG: 'wrong', NEXT_ROUND: 'next round', NEXT_INPUT: 'next input'}

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+        }, speed);
+    };
+
+     $scope.passUserInput = function(n) {
```
See if you can simplify your app logic so that you don't need a lock variable.  Simpler is easier to understand and easier to debug.  I've written stuff like if((var1 === true or var2 === false) and var3 <= 100) and paid for it with days of debugging.  Your code is good if the worst programmer in your company can easily go in and add/modify/fix something.

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+                  case 3:
+                    win.play();
+                    toggleLights(0, 5);
+                    break;
```
Maybe you can leverage the default switch case to mean the most common case which should be case 1?

ytv replied:
Good idea.  Done.

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+        resetToRound1(restart);
+    };
+
+    // resets back to round 1
```
Don't tie your function names to numerical positions.  Also would make more sense to move round = 1 into the reset method, and change resetToRound1 to something like resetGameDisplay.

ytv replied:
I see what you're saying here, but I wanted to distinguish between a complete reset where a new sequence is generated due to clicking start on turning it off and back on, and a reset back to round one (because the user messed up on strict mode) and has to start all over with the same sequence of colors.  The game wouldn't work that way if I moved round = 1 to the reset method.

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+    var sound = {1:{}, 2:{}, 3: {}, 4: {}};
+    var buzzer;
+    var win;
+
```
This isn't an init method if it depends on 'on === true'.  Maybe leverage an angular directive to help check if on is true before calling an init method if possible.

ytv replied:
Done

darrenvng commented on public/js/app.Ctrl.js:
```diff
@@ -0,0 +1,201 @@
+        }
+
+    };
+
```
Good place for a short descriptive comment of what startRound is responsible for doing.

ytv replied:
Done!

darrenvng commented on public/js/app.Service.js:
```diff
@@ -0,0 +1,90 @@
+                darkColorObj: this.getDarkColorObj(colorCode)
+            }
+    };
+
```
Definitely leverage angular directives to handle this logic for you.  Its easy in angular ngClass and ngClick to say if I do this then change this scope value, and if this scope value == something then toggle this css class.  You could get rid of most of your functions that are managing your display.  If you find a case where you need to manage some display logic yourself, don't mix it with a service that handles part of the game logic.  Move it to another /services/display.service.js file.

darrenvng commented on public/js/app.Service.js:
```diff
@@ -0,0 +1,90 @@
+        j++;
+        return result;
+    };
+
```
Your game states can be a config dependency injected where you need it so you can use it here and in your app.ctrl file.  GAME_STATE.WIN = 'win'.

ytv replied:
Done.

ytv committed:
Refactored Code

1.  Changed class names to make them more organized and understandable.  Got rid of 'row' divs
2.  Added ng-class to simplify the JS logic.  As a consequence, several new scope variables were created (i.e. green, blue, yellow, red, on, strict) and the ng-style directives were removed

ytv committed:
Refactored Code

1. Renamed several classes to more organized and understandable identifiers.
2. Removed 'row' groupings
3. Refined the positioning and font-sizes of several elements to improve responsive design

ytv committed:
Renamed from 'app.js' for consistenty for JS filenames

ytv committed:
Refactored code

1.

ytv committed:
Refactored code to improve and simplify logic

1.  Several new scope variables were created to leverage the use of ng-class.  Therefore the logic related to directly changing CSS values of elements were removed.
2.  All logic related to audio has been moved to a separate angular service in public/service/audio.service.js
3.  All variable declarations using 'var' have been changed to 'let'
4.  The default state of the game was changed from 'off' to 'on'
5.  The switch cases in scope.passUserInput() were changed from numerical values (0-3) to game states (using the const variable STATE) to improve readability

ytv committed:
Removed old, unused js files

ytv committed:
Added Rules/Instructions
